toolkit/tests/service_client/test_client.py

```diff
@@ -0,0 +1,87 @@
+from unittest.mock import AsyncMock, patch
+
+from pythonit_toolkit.api.service_client import ServiceClient
+from ward import raises, test
+
+
+@test("execute a query")
+async def _():
+    query = """
+        query{
+            users {
+                id
+            }
+        }
+    """
+    mock_response = {"data": {"users": [{"id": 1}]}}
+
+    with patch("httpx.AsyncClient.post") as post_mock:
+        post_mock.return_value = AsyncMock()
+        post_mock.return_value.json.return_value = mock_response
+
+        client = ServiceClient(
+            url="http://localhost:8050",
+            issuer="pycon",
+            audience="users-service",
+            jwt_secret="mysecret",
+        )
+
+        response = await client.execute(document=query)
+
+        assert response.data == {"users": [{"id": 1}]}
+
+
+@test("raise Exception")
+async def _():
+
+    with raises(Exception), patch("httpx.AsyncClient.post") as post_mock:
+        post_mock.return_value = AsyncMock()
+        post_mock.return_value.json.side_effect = Exception("Something went wrong")
+
+        client = ServiceClient(
+            url="http://localhost:8050",
+            issuer="pycon",
+            audience="users-service",
+            jwt_secret="mysecret",
+        )
+
+        await client.execute(document="""{ users { id }}""")
+
+
+for url, issuer, audience, jwt_secret, expected_error in [
+    ("", "pycon", "users-service", "mysecret", "Argument 'url' can't be empty"),
+    (
+        "http://localhost:8050",
+        "",
+        "users-service",
+        "mysecret",
+        "Argument 'issuer' can't be empty",
+    ),
+    (
+        "http://localhost:8050",
+        "pycon",
+        "",
+        "mysecret",
+        "Argument 'audience' can't be empty",
+    ),
+    (
+        "http://localhost:8050",
+        "pycon",
+        "users-service",
+        "",
+        "Argument 'jwt_secret' can't be empty",
+    ),
+]:
+
+    @test("raise ValueError if arguments are empty")
+    async def _():
+
+        with raises(ValueError) as exc:
```

Comment:
I don't know what to think about this :D does it even work?

Reply:
It's how ward suggests doing parametrised tests... https://ward.readthedocs.io/en/latest/guide/writing_tests.html#parameterised-testing

Reply:
They also have `each`: https://github.com/darrenburns/ward/blob/a84a4807f8b12dd1fc38da05814031095d62f6bf/tests/test_rewrite.py#L70

Maybe it will look better?

Reply:
Anyway separate test might even work for me, I know it is duplicate, but it would much clearer

Reply:
Four test for some stupid missing arguments?!? it's not too much? 
I changed to use each btw...

Reply:
You are writing tests, not business logic 🙏 tests should be clear and easy to read.

> "Four test for some stupid missing arguments"

You are not just testing missing arguments, you are testing that your client interface is correct so users of the library don't do anything weird with it/use it wrong